[Resolve #1318] Implement dump template and write output files

alexharv074 opened the pull request:
# Description

Proposed implementation of #1318 

This introduces a `dump template` command that behaves similarly to `generate` but also writes output docs to files.

The behaviour of `dump config` is changed here not to print multi-doc output to screen similar to `dump config`.

The `write` helper function is modified to optionally write to a `file_path`.

Writes files in a naming convention for `dump config` and `dump template` respectively:

```
f".dump/{stack_name}/config.yaml"
f".dump/{stack_name}/template.yaml"
```

# Usage example

## Dump config

```
% sceptre --merge-vars --dir=emr --var-file=/app/sceptre-environment/nonprod/datalake-nonprod/common-env.yaml --var-file=/app/sceptre-environment/common-env.yaml --var-file=/app/sceptre-environment/nonprod/datalake-nonprod/emr/streaming-nonprod-6100/cluster.yaml dump config emr.yaml > /dev/null
```

## Dump template

```
% sceptre --merge-vars --dir=emr --var-file=/app/sceptre-environment/nonprod/datalake-nonprod/common-env.yaml --var-file=/app/sceptre-environment/common-env.yaml --var-file=/app/sceptre-environment/nonprod/datalake-nonprod/emr/streaming-nonprod-6100/cluster.yaml dump template emr.yaml > /dev/null
```

## File structure created

```
% tree .dump/
.dump/
├── streaming-nonprod-6100-efs-pvt
│   ├── config.yaml
│   └── template.yaml
├── streaming-nonprod-6100-emr
│   ├── config.yaml
│   └── template.yaml
├── streaming-nonprod-6100-emr-profile
│   ├── config.yaml
│   └── template.yaml
└── streaming-nonprod-6100-emr-sg-pvt
    ├── config.yaml
    └── template.yaml

4 directories, 8 files
```

## File content

```yaml
% head .dump/streaming-nonprod-6100-emr/config.yaml 
---
dependencies:
- emr-profile.yaml
hooks:
  after_create:
  - !!python/object:hook.stack_termination_protection.StackTerminationProtection
    _argument: enabled
    _argument_is_resolved: false
    logger: &id001 !!python/object/apply:logging.getLogger
    - hook.stack_termination_protection
```

```yaml
% head .dump/streaming-nonprod-6100-emr/template.yaml 
---
AWSTemplateFormatVersion: '2010-09-09'

Parameters:
  MasterInstanceType:
    Type: String
  CoreInstantType:
    Type: String
  CoreInstantCount:
    Type: String
```




## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] Commit message starts with `[Resolve #issue-number]`.
- [ ] Added/Updated unit tests.
- [ ] Added/Updated integration tests (if applicable).
- [x] All unit tests (`make test`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes pre-commit validations (`pre-commit run --all-files`).
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -53,3 +55,55 @@ def dump_config(ctx, path):
+)
+@click.pass_context
+@catch_exceptions
+def dump_template(ctx, no_placeholders, path):
```
How is this different from `generate`?

alexharv074 replied:
@jfalkenstein This was what @zaro0508 and I discussed in the long thread you may not have read! The problem with `generate` always has been that it had no way to associate templates with stacks in its output for multiple stacks in a stack group, but rather just randomly orders a bunch of yaml docs that the caller has to figure out. In order to implement #1318 however we need to know which stack goes with which doc. So we agreed to deprecate `generate` and create `dump template` which behaves similarly to `dump config`.

jfalkenstein replied:
Hmmm... I'm not crazy about the idea of deprecating `generate`. Can't we just fix it to make it better?

dboitnot replied:
Yeah, I use the heck out of generate.

alexharv074 replied:
Be sure to consider all this in the context of how the new naming allows such a sane and easy to understand structure for the output templates to emerge https://github.com/alexharv074/sceptre/pull/3 

I think it is unavoidable that if we keep the existing naming (`dump config` which was named by Jon and I after a _lot_ of discussion(!) vis-a-vis `generate`) the output templates naming is going to be a dog's breakfast.

jfalkenstein commented:
I feel like there must be a better way. Is there a way we can fix the `generate` command?

zaro0508 commented on CONTRIBUTING.md:
```diff
@@ -77,11 +77,10 @@ $ git clone git@github.org:<github_username>/sceptre.git
 3. Install Sceptre for development (we recommend you use a
    [virtual environment](http://docs.python-guide.org/en/latest/dev/virtualenvs/))
 
+A convenient script is also provided:
```
no, we are going to use poetry and the poetry CLI to manage dependencies very soon, check PR #1323

alexharv074 replied:
Will remove if the other discussions are resolved.

dboitnot commented:
> I feel like there must be a better way. Is there a way we can fix the `generate` command?

Maybe a flag like `--dump` or something to the `generate` sub-command?

alexharv074 commented:
> I feel like there must be a better way. Is there a way we can fix the `generate` command?

@jfalkenstein Well if we `fixed` the generate command (we could copy this functionality there instead), we are then left with:
1. A breaking change
2. A seriously ugly inconsistency in the naming between `dump config` and `generate`. 

I think this is actually quite clean and a big improvement myself.

alexharv074 commented:
> > I feel like there must be a better way. Is there a way we can fix the `generate` command?
> 
> Maybe a flag like `--dump` or something to the `generate` sub-command?

Yes, I assume we will probably make this behaviour optional via some sort of flag in the final version.

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -53,3 +55,55 @@ def dump_config(ctx, path):
+
+    output_format = "json" if context.output_format == "json" else "yaml"
+
+    if len(output) == 1:
+        write(output[0][stack.external_name], output_format)
+    else:
+        for template in output:
+            write(template, output_format)
```
It strikes me as odd that the output format is fundamentally different if it's just one stack vs. if it's numerous stacks.

alexharv074 replied:
It makes sense if you use it, and, again, is what you and I agreed on already after lengthy discussions for `dump config`. In the case of a single stack—most of the time—you don't want to see the stack names in this output; you just want the yaml output. In the case of a stack group however, you need to know which stack goes with which doc.

HOWEVER, please note that in the implementation of Khai's feature, I have changed this back to multi-doc yaml output without the stack names. That's because the .dump structure that is created allows us to know which stack goes with which doc.

alexharv074 replied:
Closing this because it's not relevant after I merged Part 2 into here.

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -53,3 +55,55 @@ def dump_config(ctx, path):
+        if template is None:
+            logger.warning(f"{stack.external_name} does not exist")
+        else:
+            output.append({stack.external_name: template})
```
Why a list of dicts and not just one big dict?

alexharv074 replied:
Again. For consistency with `dump config`. I can't remember to be honest why we did it this way.

jfalkenstein replied:
That feels a bit odd to me

alexharv074 replied:
I _think_ it may be to work around or hack the limitations of the `write` helper. If we need to change `dump config`, perhaps this should be looked at in a separate issue/PR?

jfalkenstein replied:
My only concern is this: Let's say I need to pipe a template over to a linter or awscli for some reason because it has jinja and other stuff in it. I want to do something like: `sceptre dump template my/sceptre/config.yaml | cfn-lint -`. When you're outputting the templates inside a larger data structure like this, that's not a valid template. 

I understand your concern of wanting to be able to associate the template with a stack, so there are a few options I can think of:

Option 1: Rather than this, you could output the stack name via a logger rather than echoing it. That will result it going to stderr rather than stdout. That way, if you pipe output to another command or use `>`, it won't actually get included. So you could alternate by logging the stack name and then writing the output.

Option 2: Have a `--stack-names` flag that optionally makes it output this way but by default it does not.

alexharv074 replied:
@jfalkenstein you are misunderstanding the implementation and behaviour. Here's a demo, showing a yaml stream that can be parsed by external tools is always emitted, even in the case of multiple stacks:

<img width="1770" alt="Screen Shot 2023-04-21 at 8 20 03 pm" src="https://user-images.githubusercontent.com/2305330/233612735-de6664aa-b83c-4bea-8a13-4bcbe6e33328.png">

jfalkenstein commented on sceptre/plan/actions.py:
```diff
@@ -1152,3 +1152,10 @@ def dump_config(self, config_reader: ConfigReader):
         stack_path = normalise_path(self.stack.name + ".yaml")
         return config_reader.read(stack_path)
+
+    @add_stack_hooks
+    def dump_template(self):
+        """
+        Returns the Template for the Stack
+        """
+        return self.stack.template.body
```
This is identical to `generate()`.

alexharv074 replied:
Yes, it needs to be identical to `generate()` since the latter is deprecated and would be removed. Also, the naming standard requires that CLI commands have a method named after the CLI command (at least I hope we preserve and continue that naming standard!).

alexharv074 replied:
If we agree to alias `generate` and `dump template` of course then perhaps we only need one method here.

jfalkenstein replied:
Yeah, I would say that you should make `generate` here call this method and return its result. That should be a pretty harmless way to start aliasing.

alexharv074 replied:
ok done

jfalkenstein replied:
If you're deprecating `generate`, you definitely should do that correctly, both in docs and using the `deprecated` decorator.

alexharv074 replied:
ok done

jfalkenstein commented on venv.sh:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
```
This file doesn't belong in this PR.

jfalkenstein reviewed (commented):
I think I'm just very concerned about adding a completely new command that is ALMOST identical to the old command except it outputs in a slightly different format. I really don't think this is the right way to proceed.

alexharv074 commented:
> I think I'm just very concerned about adding a completely new command that is ALMOST identical to the old command except it outputs in a slightly different format. I really don't think this is the right way to proceed.

I guess you would need to say then what the right way to proceed is then, because I can't personally see any other option. The legacy `sceptre generate` feature is a mess though. It always has been very broken that it doesn't tell you the stack name and dumps a randomly ordered list of yaml docs.

alexharv074 commented on sceptre/cli/helpers.py:
```diff
@@ -78,6 +78,8 @@ def write(var, output_format="json", no_colour=True):
     :param no_colour: Whether to colour stack statuses
     :type no_colour: bool
+    :param file_path: Optional path to a file to save the output
+    :type file_path: str, optional
```
Yes these are Pathlib paths not strings. Will fix.

alexharv074 replied:
fixed

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -48,8 +52,70 @@ def dump_config(ctx, path):
+
     if len(output) == 1:
-        write(output[0][stack.external_name], output_format)
+        process_config(output[0])
     else:
         for config in output:
-            write(config, output_format)
+            process_config(config)
```
I don't see a reason any more to have this `if/else` here. I think you could just have the for loop.

jfalkenstein replied:
`process_config(output[0])` is the same as doing this: 

```python
for config in output[:1]:
    process_config(output)
```

And if output is a list of only one item, it's the same as doing this, because it'll only loop once:

```python
for config in output:
    process_config(output)
```

Because remember, looping through a list like this is essentially the same as doing:

```python
for i in range(len(output)):
    process_config(output[i])
```

But if that's the flow you want for >1 as well, you can just use that for output, no matter if there are 0 items (it will loop zero times), 1 items (it will loop once), or 100 items. 

Anyway, it's all the same. You don't need the if/else.

alexharv074 replied:
@jfalkenstein ah! I see your point. They used to do different things and now they don't. Updated.

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -48,8 +52,70 @@ def dump_config(ctx, path):
+            process_config(config)
+
+
+@dump_group.command(name="template")
```
I think I disagree. Sometimes (maybe even most of the time) we'd want to dump the template. But sometimes we'll want to dump the stack-config. 

But I do agree that we shouldn't write (at least by default) both to files and console. I think (in line with every other Sceptre command), we should default to console output. However, you could use `--file` to output to file.

alexharv074 replied:
I disagree strongly. I never, ever want both, and a change like this would seriously break all our tooling too.

alexharv074 replied:
So, on thinking about the second part, the problem I see is this. If the command does not always dump to the screen _and_ a file, then we arrive back at the problem that the default behaviour of printing to the screen should also by defaulting tell the user about the stack name. If I do that, then @jfalkenstein 's wish that we can simply redirect a multi-doc yaml to a file is not satisfied. I also like the output better this way, and the dumped files can tell me which template goes with which stack. 

Thoughts?

jfalkenstein replied:
I don't think we ought to dump to a file by default. If there's a `--file` option, though... I don't really care if it goes to the console. I think it would be cleaner if what goes to the console was just "Dumped contents were output to {file path}", though.

dboitnot replied:
I agree that console output should be the default and I agree that a multi-doc YAML output makes sense. 

An alternative would be to pipe the rendered content to STDIN and meta information like the stack name to STDERR.

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -48,8 +52,70 @@ def dump_config(ctx, path):
+
+    output = []
+    for stack, template in responses.items():
+        if template is None:
+            logger.warning(f"{stack.external_name} does not exist")
```
I don't understand the case where `template` would be `None`...

alexharv074 replied:
As hinted in the following line, a `None` would be returned in the case of the stack which hasn't been created or for some reason doesn't exist. I have a vague memory but probably can't remember how to actually reproduce the scenario that led to the if statement.

jfalkenstein replied:
That's not true. You can absolutely generate a template when a stack isn't created. That's how the diff command works when there's no stack deployed yet.

alexharv074 replied:
Ah you know what, it came from the `fetch_remote_template` command. I may have copy/pasted that into the dump config command in error. Ok, I've removed it in both `dump config` and `dump template`.

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -48,8 +52,70 @@ def dump_config(ctx, path):
+    "--no-placeholders",
+    is_flag=True,
+    help="If True, no placeholder values will be supplied for resolvers that cannot be resolved.",
+)
```
I think there ought to be an option, like `--only` that will exclusively output the template at the indicated path and not include any other configs passed along. Probably the same for `dump config` as well.

alexharv074 replied:
Yes possibly, but could we do that in a separate change later?

jfalkenstein replied:
That's a pretty simple change to make, IMO. But yes, it could be done later.

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -48,8 +52,70 @@ def dump_config(ctx, path):
+        file_path = Path(f".dump/{stack_name}/template.yaml")
+        write(template[stack_name], output_format, no_colour=True, file_path=file_path)
+
+    if len(output) == 1:
+        process_template(output[0])
+    else:
+        for template in output:
+            process_template(template)
```
Same comment here... I think we could just have the for loop.

alexharv074 replied:
Same response here. Unless I'm dreaming these two branches do different things:

```python
    if len(output) == 1:
        process_template(output[0])
    else:
        for template in output:
            process_template(template)                                                                                                                                                                                            
```

jfalkenstein replied:
You're indeed dreaming!

jfalkenstein commented on sceptre/cli/helpers.py:
```diff
@@ -93,6 +95,13 @@ def write(var, output_format="json", no_colour=True):
     click.echo(output)
 
+    if file_path:
+        dir_path = file_path.parent
```
You're still treating a string like a pathlib.Path object here.

alexharv074 replied:
Sorry, can you elaborate? I am not seeing the problem there.

alexharv074 replied:
I've updated the type hints to clarify that a pathlib.Path is expected here.

jfalkenstein commented on sceptre/cli/helpers.py:
```diff
@@ -93,6 +95,13 @@ def write(var, output_format="json", no_colour=True):
 
     click.echo(output)
```
Echo has a `file` argument that you can use to write the it to a file. https://click.palletsprojects.com/en/8.1.x/api/#click.echo. It defaults to stdout. So you could basically do:
* If `file_path` is passed, make the `file` argument a file object at that path
* If it is not passed, make the `file` argument `sys.stdout`

alexharv074 replied:
Ok, I see the improvement in that, done.

jfalkenstein commented on sceptre/plan/plan.py:
```diff
@@ -440,3 +440,13 @@ def dump_config(self, *args):
+        :returns: A dictionary of Stacks and their template body.
+        :rtype: dict
+        """
+        self.resolve(command=self.generate.__name__)
```
Why is this calling `generate`? I think that might be a mistake.

alexharv074 replied:
fixed.

jfalkenstein reviewed (changes requested):
I've left some comments. We'll still need to update docs and properly deprecate `generate`.

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -48,8 +52,70 @@ def dump_config(ctx, path):
+
+    def process_template(template):
+        stack_name = list(template.keys())[0]
+        file_path = Path(f".dump/{stack_name}/template.yaml")
```
@alexharv074 , to my point elsewhere, it seems you might not realize that logs are configured to go to stderr in Sceptre (it's actually the default for Python), so logs wouldn't interfere with using `>` because that sends stdout. They would only get sucked up if you used `$>`.

jfalkenstein replied:
Also, I might not be right about this, but I suspect this path isn't Windows-safe. You could instead do one of the following to guarantee it works with Windows:
```python
Path(".dump", stack_name, "template.yaml")
# or
Path(".dump") / stack_name / "template.yaml")

jfalkenstein replied:
Also, the template might be json, not yaml.

alexharv074 replied:
Ok fixed fixed and fixed.

jfalkenstein commented on sceptre/cli/helpers.py:
```diff
@@ -173,6 +189,23 @@ def _generate_text(stream):
     return stream
 
 
+def process_template(
```
While `process_template` as a nested function inside your other functions was a perfectly acceptable name, since you would contextually understand what "processing" meant. However, since you've extracted this function out and put it in `helpers`, it isn't obvious at all what "processing" means. I think we should have a better name on this function that clearly describes what it is supposed to do. Maybe `dump_dict_to_file`?

alexharv074 replied:
fixed

jfalkenstein commented on sceptre/cli/helpers.py:
```diff
@@ -173,6 +189,23 @@ def _generate_text(stream):
 
 
+def process_template(
+    template: dict, output_format: str, type: str = "template"
```
It's generally discouraged to shadow builtins like `type`.

alexharv074 replied:
fixed

jfalkenstein commented on sceptre/cli/helpers.py:
```diff
@@ -173,6 +189,23 @@ def _generate_text(stream):
+    """
+    stack_name = list(template.keys())[0]
+    file_path = Path(".dump") / stack_name / f"{type}.{output_format}"
+    logger.info(f"{stack_name} dumping {type} to {file_path}")
+    write(template[stack_name], output_format, no_colour=True, file_path=file_path)
+    logger.info(f"{stack_name} done.")
```
I don't think we need both of these log messages, at least not both at the `info` level. 

I think having the initial message would be very helpful in the case that writing fails for some reason (not sure why that would be the case, though...). So I think the first log message should be emitted at the `debug` level rather than `info` and the second log message should include the file_path as well and be something like: `f'{stack_name} dumped to file at {file_path}'`.

alexharv074 replied:
@jfalkenstein the message is proposed by @zaro0508. I would add that without a second logger message after dump has completed, the first will be probably not noticed.

alexharv074 replied:
have a look now.

jfalkenstein commented on sceptre/plan/actions.py:
```diff
@@ -624,12 +628,13 @@ def _convert_to_url(self, summaries):
 
         return new_summaries
 
+    @deprecation.deprecated("4.0.0", "6.0.0", __version__, "Use dump template instead.")
     @add_stack_hooks
```
Also, add a `fail_if_not_removed` test on this as well.

jfalkenstein commented on sceptre/plan/plan.py:
```diff
@@ -440,3 +440,13 @@ def dump_config(self, *args):
         """
         self.resolve(command=self.dump_config.__name__)
         return self._execute(self.config_reader, *args)
+
```
You ought to deprecate `generate` above as well.

alexharv074 replied:
done

alexharv074 commented on sceptre/cli/helpers.py:
```diff
@@ -173,6 +189,23 @@ def _generate_text(stream):
+    """
+    Helper to write templates and configs used in dump commands.
+
+    :param template: the template.
```
You'll have to elaborate ....

jfalkenstein replied:
Your `template` variable here isn't actually a template. In other words, your variable naming lies. For a helper function that's meant to stand on its own, the interface must be clear, simple, and functional. But you're making a lot of assumptions here about what `template` is. 

It should take a `stack_name` and a `template_dict` parameter instead of a dict that happens to have a key that is the stack name and a value that is the actual template dict. Your code would be simpler, more readable, and easier to interact with if you did that.

alexharv074 replied:
I'm not convinced that a variable named `template_dict` of type `dict` is clearer than `template: dict` and to my mind the whole comes out more rather than less messy afterwards. Perhaps we are better off without this helper at all and I revert to how it was before?

alexharv074 replied:
For example on the caller side:

```python
    for template in output:
        dump_to_file(list(template.keys())[0], template, output_format)
```

We don't want duplication of `list(template.keys())[0]` all over the place for the sake helping the reader of that helper function since most people are going to be reading the CLI functions; not the helpers.

jfalkenstein replied:
My point is that you're calling the dict a "template", but it's actually a dict with a single key indicating the stack name and a its value is the actual template. In other words, your variable name is wrong and the interface of the function is more challenging than it ought to be. 

Rather than assuming this arbitrary structure that is a mouthful to describe, I'm just saying take two parameters instead of this complex dict structure that is isn't well-typed.

jfalkenstein reviewed (dismissed):
I know you won't like this, but `generate` has integration tests and I believe that we should have integration tests on these dump commands as well.

alexharv074 commented:
> I know you won't like this, but `generate` has integration tests and I believe that we should have integration tests on these dump commands as well.

I've refactored the `generate` test to instead call `dump_template`. As for the other dump commands, they are on a different branch so will have to go in different PRs and this one needs to be merged first.

alexharv074 commented:
@jfalkenstein  have a look now. Here's what I changed:
- implemented `--to-file` making the default behaviour not to create the dump files.
- refactored the process files logic you didn't like moving it all back the dump commands.

jfalkenstein commented on sceptre/cli/dump.py:
```diff
@@ -18,9 +22,12 @@ def dump_group():
 @dump_group.command(name="config")
 @click.argument("path")
+@click.option(
+    "--to-file", is_flag=True, help="If True, also dump the template to a local file."
```
Never mind. I see what we're doing now.

zaro0508 reviewed (changes requested):
`generate` is found in a few locations in the docs.  if we are deprecating that command in this same PR then could we also update to dump in the docs as well? 

I think it might be cleaner to deprecate the generate command in a follow on PR.

alexharv074 commented on sceptre/cli/dump.py:
```diff
@@ -39,17 +46,91 @@ def dump_config(ctx, path):
+
+@dump_group.command(name="all")
+@click.argument("path")
+@click.option(
```
not intentional, fixed.

alexharv074 commented:
> `generate` is found in a few locations in the docs. if we are deprecating that command in this same PR then could we also update to dump in the docs as well?
> 
> I think it might be cleaner to deprecate the generate command in a follow on PR.

A few hits on the word when I `grep -rw generate`, however those are not hits on the subcommand. Are you sure?

alexharv074 commented on sceptre/cli/helpers.py:
```diff
@@ -93,6 +102,13 @@ def write(var, output_format="json", no_colour=True):
 
     click.echo(output)
```
Ok how about now? (I'm not sure whether people want the text to go to both the screen and file if file_path is specified).

zaro0508 commented:
> A few hits on the word when I grep -rw generate, however those are not hits on the subcommand. Are you sure?

This is what i see 

```
docs/_source/docs/hooks.rst:``before_generate`` or ``after_generate`` - run hook before or after generating stack template.
docs/_source/docs/resolvers.rst:  ``!stack_output other_stack.yaml::OutputName`` and you run the ``dump template`` command, the generated
docs/_source/docs/templates.rst:to generate CloudFormation Template as a `json` string.
docs/_source/docs/templates.rst:  To generate templates using Troposphere you must install the
docs/_source/docs/templates.rst:AWS CDK can be used to programmatically generate templates for your stacks and then Sceptre can
```
I think those references should be changed to the `dump` command


Also didn't we decide to redirect the generate to alias the dump command?  if so it looks like it wasn't done in template.py
```
sceptre/cli/template.py:@click.command(name="generate", short_help="Prints the template.")
sceptre/cli/template.py:def generate_command(ctx, no_placeholders, path):
```

alexharv074 commented:
> This is what i see
> 
> ```
> docs/_source/docs/hooks.rst:``before_generate`` or ``after_generate`` - run hook before or after generating stack template.
> docs/_source/docs/resolvers.rst:  ``!stack_output other_stack.yaml::OutputName`` and you run the ``dump template`` command, the generated
> docs/_source/docs/templates.rst:to generate CloudFormation Template as a `json` string.
> docs/_source/docs/templates.rst:  To generate templates using Troposphere you must install the
> docs/_source/docs/templates.rst:AWS CDK can be used to programmatically generate templates for your stacks and then Sceptre can
> ```
> 
> I think those references should be changed to the `dump` command

@zaro0508 I didn't actually know about the `before_generate` and `after_generate` hooks but I don't see how those can be renamed. `before_dump_template` is both an ugly name and does not make sense in that context. It would have to stay as `before_generate` I think. Remember, it doesn't mean "before you run the generate command" ; it means "before Sceptre generates its template". Generating the template internally, and dumping it out for the caller to look at, are different things.

alexharv074 commented:
> Also didn't we decide to alias the generate to the dump command? if so, it looks like it wasn't done in template.py
> 
> ```
> sceptre/cli/template.py:@click.command(name="generate", short_help="Prints the template.")
> sceptre/cli/template.py:def generate_command(ctx, no_placeholders, path):
> ```

Yes but it was aliased not in the CLI but in the API which is I think the way it should be.

zaro0508 commented:
> Yes but it was aliased not in the CLI but in the API which is I think the way it should be.

What is your reasoning for not aliasing it in the CLI as well?

alexharv074 commented:
> > Yes but it was aliased not in the CLI but in the API which is I think the way it should be.
> 
> What is your reasoning for not aliasing it in the CLI as well?

Well, aliasing in both places will just result in unused code paths. If I point the generate CLI command at the dump_template API, why not just remove the generate API since it is no longer used? 

In the end I don't really mind how it is done or necessarily see the point in aliasing this. The implementation above has the advantage that it seems to be the one that @jfalkenstein agreed to and not to mention that it's the one that I already implemented!
